TypeCobol/Compiler/CodeModel/SymbolTable.cs

```diff
@@ -57,40 +31,63 @@ public SymbolTable(SymbolTable enclosing, Scope current)
-        [NotNull]
-        private List<T> GetFromTableAndEnclosing<T>(string head,
-            Func<SymbolTable, IDictionary<string, List<T>>> getTableFunction, Scope maxScope = Scope.Intrinsic) where T : Node
+        private delegate void MatchDelegate<TNode>(List<TNode> found,
+            in TNode head,
+            in QualifiedName name,
+            int nameIndex,
+            in TNode current)
+            where TNode : Node;
+
+        private List<T> GetFromTableAndEnclosing<T>(QualifiedName name,
+            Func<SymbolTable, IDictionary<string, List<T>>> getEntries,
+            MatchDelegate<T> match,
+            Scope maxScope,
+            Scope? requiredScope = null)
+            where T : Node
         {
-            System.Diagnostics.Debug.Assert(head != null);
+            System.Diagnostics.Debug.Assert(requiredScope == null || requiredScope <= maxScope);
+
             var result = new List<T>();
-            this.GetFromTableAndEnclosing2(head, getTableFunction, result, maxScope);
-            
+            bool skipParentGlobal = false;
+
+            // Iterate over scopes from current up to maxScope
+            var currentSymbolTable = this;
+            while (currentSymbolTable != null && currentSymbolTable.CurrentScope >= maxScope)
+            {
+                // Retrieve candidates from current SymbolTable
+                if (!getEntries(currentSymbolTable).TryGetValue(name.Head, out var candidates)) candidates = new List<T>();
+
+                // We don't have to perform matching if we already have found a result (except if current scope is required)
+                if (!skipParentGlobal || currentSymbolTable.CurrentScope == requiredScope)
+                {
+                    // Proceed to matching
+                    foreach (var candidate in candidates)
+                    {
+                        match(result, candidate, name, name.Count - 1, candidate);
+                    }
+
+                    // Skip Global SymbolTable from parent if we already have found a match
+                    skipParentGlobal = currentSymbolTable.CurrentScope == Scope.Global && result.Count > 0;
+                }
+
+                // Jump to parent SymbolTable
+                currentSymbolTable = currentSymbolTable.EnclosingScope;
+            }
+
             return result;
         }
 
-        private void GetFromTableAndEnclosing2<T>([NotNull] string head,
-           Func<SymbolTable, IDictionary<string, List<T>>> getTableFunction, List<T> result, Scope maxScope = Scope.Intrinsic) where T : Node
+        private static void MatchUsingName<T>(List<T> found, in T head, in QualifiedName name, int nameIndex, in T current)
+            where T : Node
         {
-            var table = getTableFunction.Invoke(this);
-            GetFromTable(head, table, result);
-            if (EnclosingScope != null && EnclosingScope.CurrentScope >= maxScope)
+            if (Match(head.QualifiedName, name))
```

Comment:
`QualifiedName` is a very CPU consuming property.
Can you use property `Name` and `Parent` only ?

Reply:
Method rewritten in d957623bdbec59c56a78adeea8c50f0115f41556